# ✈️ Airline Operations & Disruption Intelligence — Business Insights

**Beginner-friendly Business Insights Notebook**

### Purpose
Convert **EDA + Statistics** into simple business findings and recommendations.

### Format
**Finding → Evidence → Business impact → Recommendation**

> This notebook does not add new statistical tests. It focuses on explaining the results in business language.


## 1. Project flow

**Business Problem → Data Cleaning → Airport Reference → Weather Integration → EDA → Statistics → Business Insights → Power BI**

The technical analysis is already completed. This stage is about communicating what the analysis means for airline operations.


## 2. Import libraries

In [3]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)


## 3. Load the final enriched dataset

This file was created after:

**BTS cleaning → airport reference → Open-Meteo weather integration**


In [4]:
file_path = r"D:\Data Analyst\EXCEL\airline-operations-intelligence\Airline-Operations-Intelligence\data\processed\flights_2026_q1.csv"

df = pd.read_csv(file_path, low_memory=False)

print("Data loaded successfully")
print("Rows:", f"{len(df):,}")
print("Columns:", len(df.columns))


Data loaded successfully
Rows: 1,847,242
Columns: 78


In [5]:
df.head()

,year,quarter,month,day_of_month,day_of_week,fl_date,mkt_unique_carrier,branded_code_share,mkt_carrier_airline_id,mkt_carrier,mkt_carrier_fl_num,origin,origin_city_name,origin_state_abr,origin_state_nm,dest,dest_city_name,dest_state_abr,dest_state_nm,crs_dep_time,dep_time,dep_delay,dep_delay_new,dep_del15,taxi_out,taxi_in,crs_arr_time,arr_time,arr_delay,arr_delay_new,arr_del15,cancelled,cancellation_code,diverted,crs_elapsed_time,actual_elapsed_time,air_time,distance,carrier_delay,weather_delay,nas_delay,security_delay,late_aircraft_delay,flight_status,dep_time_at_cancellation,missing_operational_duration,extreme_dep_delay,extreme_arr_delay,is_cancelled,is_diverted,is_irregular_operation,is_departure_delayed,is_arrival_delayed,severe_departure_delay,severe_arrival_delay,origin_airport_name,origin_city,origin_country,origin_latitude,origin_longitude,origin_timezone,dest_airport_name,dest_city,dest_country,dest_latitude,dest_longitude,dest_timezone,dep_hour,dep_minute,scheduled_departure,departure_hour,temperature,precipitation,wind_speed,weather_code,weather_matched,weather_category,bad_weather
0,2026,1,1,1,4,2026-01-01,AA,AA,19805,AA,1,JFK,"New York, NY",NY,New York,LAX,"Los Angeles, CA",CA,California,700,657.0,-3.0,0.0,0.0,67.0,11.0,1021,1104.0,43.0,43.0,1.0,0.0,NaN,0.0,381.0,427.0,349.0,2475.0,0.0,0.0,43.0,0.0,0.0,COMPLETED,NaN,False,False,False,0,0,0,0,1,0,0,John F Kennedy International Airport,New York,United States,40.639801,-73.778900,-5,Los Angeles International Airport,Los Angeles,United States,33.942501,-118.407997,-8,7,0,2026-01-01 07:00:00,7,0.9,0.0,18.3,3,True,Cloudy,0
1,2026,1,1,1,4,2026-01-01,AA,AA,19805,AA,10,LAX,"Los Angeles, CA",CA,California,JFK,"New York, NY",NY,New York,2130,2128.0,-2.0,0.0,0.0,16.0,6.0,551,520.0,-31.0,0.0,0.0,0.0,NaN,0.0,321.0,292.0,270.0,2475.0,NaN,NaN,NaN,NaN,NaN,COMPLETED,NaN,False,False,False,0,0,0,0,0,0,0,Los Angeles International Airport,Los Angeles,United States,33.942501,-118.407997,-8,John F Kennedy International Airport,New York,United States,40.639801,-73.778900,-5,21,30,2026-01-01 21:30:00,21,17.3,0.2,4.5,51,True,Rain,1
2,2026,1,1,1,4,2026-01-01,AA,AA,19805,AA,1002,MIA,"Miami, FL",FL,Florida,MSY,"New Orleans, LA",LA,Louisiana,2245,2252.0,7.0,7.0,0.0,16.0,3.0,2359,2356.0,-3.0,0.0,0.0,0.0,NaN,0.0,134.0,124.0,105.0,674.0,NaN,NaN,NaN,NaN,NaN,COMPLETED,NaN,False,False,False,0,0,0,0,0,0,0,Miami International Airport,Miami,United States,25.793200,-80.290604,-5,Louis Armstrong New Orleans International Airport,New Orleans,United States,29.993401,-90.258003,-6,22,45,2026-01-01 22:45:00,22,18.2,0.0,7.9,0,True,Clear,0
3,2026,1,1,1,4,2026-01-01,AA,AA,19805,AA,1003,DEN,"Denver, CO",CO,Colorado,MIA,"Miami, FL",FL,Florida,2338,2338.0,0.0,0.0,0.0,13.0,6.0,527,508.0,-19.0,0.0,0.0,0.0,NaN,0.0,229.0,210.0,191.0,1709.0,NaN,NaN,NaN,NaN,NaN,COMPLETED,NaN,False,False,False,0,0,0,0,0,0,0,Denver International Airport,Denver,United States,39.861698,-104.672997,-7,Miami International Airport,Miami,United States,25.793200,-80.290604,-5,23,38,2026-01-01 23:38:00,23,12.5,0.0,4.8,2,True,Cloudy,0
4,2026,1,1,1,4,2026-01-01,AA,AA,19805,AA,1004,BOS,"Boston, MA",MA,Massachusetts,CLT,"Charlotte, NC",NC,North Carolina,757,756.0,-1.0,0.0,0.0,26.0,4.0,1030,1023.0,-7.0,0.0,0.0,0.0,NaN,0.0,153.0,147.0,117.0,728.0,NaN,NaN,NaN,NaN,NaN,COMPLETED,NaN,False,False,False,0,0,0,0,0,0,0,General Edward Lawrence Logan International Ai...,Boston,United States,42.364300,-71.005203,-5,Charlotte Douglas International Airport,Charlotte,United States,35.214001,-80.943100,-5,7,57,2026-01-01 07:57:00,7,-2.4,0.1,16.7,71,True,Snow,1


## 4. Overall operational performance

### Business question
**What is the overall operational condition of the flights?**


In [6]:
total_flights = len(df)
cancelled_flights = df["cancelled"].sum()
diverted_flights = df["diverted"].sum()

completed_df = df[
    (df["cancelled"] == 0) &
    (df["diverted"] == 0)
]

delay_rate = completed_df["arr_del15"].mean() * 100
cancellation_rate = cancelled_flights / total_flights * 100
diversion_rate = diverted_flights / total_flights * 100

print("Total flights:", f"{total_flights:,}")
print("Completed flights:", f"{len(completed_df):,}")
print("Arrival delay rate:", round(delay_rate, 2), "%")
print("Cancellation rate:", round(cancellation_rate, 2), "%")
print("Diversion rate:", round(diversion_rate, 2), "%")


Total flights: 1,847,242
Completed flights: 1,780,194
Arrival delay rate: 21.59 %
Cancellation rate: 3.37 %
Diversion rate: 0.26 %


### Business interpretation

> During the study period, the dataset contained **[18,47,242] flights**. The arrival delay rate was **[21.59]%**, cancellation rate was **[3.37]%**, and diversion rate was **[0.26]%**.

This gives management the starting point for understanding operational performance.


## 5. Delay performance

### Business question
**How large are the delays?**


In [7]:
avg_departure_delay = completed_df["dep_delay"].mean()
median_departure_delay = completed_df["dep_delay"].median()
avg_arrival_delay = completed_df["arr_delay"].mean()
median_arrival_delay = completed_df["arr_delay"].median()

print("Average departure delay:", round(avg_departure_delay, 2), "minutes")
print("Median departure delay:", round(median_departure_delay, 2), "minutes")
print("Average arrival delay:", round(avg_arrival_delay, 2), "minutes")
print("Median arrival delay:", round(median_arrival_delay, 2), "minutes")


Average departure delay: 13.81 minutes
Median departure delay: -2.0 minutes
Average arrival delay: 7.25 minutes
Median arrival delay: -7.0 minutes


### Business interpretation

If the average is much higher than the median, some severe delays are increasing the average.

**Business impact:** severe delays can create downstream disruption for later flights and passengers.


## 6. Monthly operational performance

### Business question
**Is operational performance changing over time?**


In [8]:
df["fl_date"] = pd.to_datetime(df["fl_date"])

monthly_summary = (
    df.groupby(df["fl_date"].dt.to_period("M"))
    .agg(
        flights=("fl_date", "size"),
        cancelled=("cancelled", "sum"),
        diverted=("diverted", "sum")
    )
)

monthly_summary["cancellation_rate"] = monthly_summary["cancelled"] / monthly_summary["flights"] * 100
monthly_summary["diversion_rate"] = monthly_summary["diverted"] / monthly_summary["flights"] * 100

monthly_summary


,flights,cancelled,diverted,cancellation_rate,diversion_rate
fl_date,,,,,
2026-01,602953,29078.0,1276.0,4.822598,0.211625
2026-02,568789,13348.0,1144.0,2.346740,0.201129
2026-03,675500,19885.0,2317.0,2.943745,0.343005


## 7. Airline performance

### Business question
**Which airlines show higher operational risk?**

Use a minimum flight count so small airlines do not dominate the ranking.


In [9]:
airline_summary = (
    df.groupby("mkt_carrier")
    .agg(
        flights=("mkt_carrier", "size"),
        avg_arrival_delay=("arr_delay", "mean"),
        delayed_flights=("arr_del15", "sum"),
        cancelled=("cancelled", "sum"),
        diverted=("diverted", "sum")
    )
)

airline_summary["delay_rate"] = df.groupby("mkt_carrier")["arr_del15"].mean() * 100
airline_summary["cancellation_rate"] = airline_summary["cancelled"] / airline_summary["flights"] * 100
airline_summary["diversion_rate"] = airline_summary["diverted"] / airline_summary["flights"] * 100

airline_summary = airline_summary[airline_summary["flights"] >= 100]
airline_summary.sort_values("delay_rate", ascending=False).head(10)


,flights,avg_arrival_delay,delayed_flights,cancelled,diverted,delay_rate,cancellation_rate,diversion_rate
mkt_carrier,,,,,,,,
NK,36379,19.661396,11761.0,3123.0,55.0,35.423632,8.584623,0.151186
B6,57320,15.625995,16901.0,3010.0,153.0,31.207415,5.251221,0.266923
F9,50557,13.014408,13128.0,1405.0,81.0,26.753072,2.779041,0.160215
G4,30572,13.490645,7475.0,479.0,108.0,24.929131,1.566793,0.353264
AA,483902,10.322157,105874.0,22945.0,1492.0,23.042887,4.741663,0.308327
DL,386832,5.853817,74823.0,14328.0,840.0,20.131893,3.703933,0.217149
WN,327083,3.200954,64390.0,5659.0,624.0,20.071696,1.730142,0.190777
UA,358108,6.235863,68900.0,9204.0,998.0,19.804200,2.570174,0.278687
AS,116489,2.175787,21174.0,2158.0,386.0,18.582650,1.852535,0.331362


### Business interpretation

Airlines at the top have higher observed delay rates among airlines with sufficient flight volume.

**Recommendation:** investigate airport mix, route mix, schedule timing, and recorded delay causes before deciding what is driving the difference.


## 8. Airport performance

### Business question
**Where are operational problems concentrated?**


In [10]:
airport_summary = (
    df.groupby("origin")
    .agg(
        flights=("origin", "size"),
        avg_arrival_delay=("arr_delay", "mean"),
        delayed_flights=("arr_del15", "sum"),
        cancelled=("cancelled", "sum"),
        diverted=("diverted", "sum")
    )
)

airport_summary["delay_rate"] = df.groupby("origin")["arr_del15"].mean() * 100
airport_summary["cancellation_rate"] = airport_summary["cancelled"] / airport_summary["flights"] * 100

airport_summary = airport_summary[airport_summary["flights"] >= 1000]
airport_summary.sort_values("avg_arrival_delay", ascending=False).head(10)


,flights,avg_arrival_delay,delayed_flights,cancelled,diverted,delay_rate,cancellation_rate
origin,,,,,,,
SBN,2043,31.841837,687.0,78.0,5.0,35.051020,3.817915
TVC,1070,25.356713,337.0,71.0,1.0,33.767535,6.635514
CAK,1138,23.560150,326.0,73.0,1.0,30.639098,6.414763
FWA,1845,22.130337,507.0,64.0,1.0,28.483146,3.468835
ASE,3237,19.483827,840.0,324.0,7.0,28.905712,10.009268
FLL,26789,19.401207,8754.0,1199.0,62.0,34.291758,4.475718
ATW,2284,19.303489,606.0,103.0,3.0,27.823691,4.509632
GRB,1472,19.301282,424.0,62.0,6.0,30.199430,4.211957
CID,2401,18.242869,588.0,82.0,5.0,25.410545,3.415244


### Business interpretation

High-delay airports are operational hotspots.

**Recommendation:** investigate these airports using departure hour, airline, route, delay cause, and weather information.


## 9. Route performance

### Business question
**Are delays concentrated on specific routes?**


In [11]:
df["route"] = df["origin"] + " → " + df["dest"]

route_summary = (
    df.groupby("route")
    .agg(
        flights=("route", "size"),
        avg_arrival_delay=("arr_delay", "mean"),
        delay_rate=("arr_del15", "mean")
    )
)

route_summary["delay_rate"] *= 100
route_summary = route_summary[route_summary["flights"] >= 500]

route_summary.sort_values("delay_rate", ascending=False).head(10)


,flights,avg_arrival_delay,delay_rate
route,,,
CVG → DCA,508,37.898947,46.736842
JFK → DCA,707,28.229938,44.290123
FLL → SJU,624,25.689256,43.140496
ORD → ASE,511,32.535484,43.010753
ORD → SBN,714,31.162963,42.518519
PHL → SJU,571,21.138837,41.651032
ORD → MKE,1233,33.720651,41.645244
DTW → GRR,515,39.736220,41.141732
BOS → PBI,694,25.194785,40.950920


### Business interpretation

High-delay routes should be investigated using airport, airline, schedule, and delay-cause information.

The minimum-flight rule prevents a route with very few flights from producing a misleading ranking.


## 10. Delay causes

### Business question
**What contributes the most recorded delay minutes?**


In [12]:
delay_columns = {
    "Carrier Delay": "carrier_delay",
    "Weather Delay": "weather_delay",
    "NAS Delay": "nas_delay",
    "Security Delay": "security_delay",
    "Late Aircraft Delay": "late_aircraft_delay"
}

delay_totals = {}

for cause, column in delay_columns.items():
    if column in df.columns:
        delay_totals[cause] = df[column].fillna(0).sum()

delay_totals = pd.Series(delay_totals).sort_values(ascending=False)

print(delay_totals)
print("Largest recorded delay cause:", delay_totals.idxmax())


Late Aircraft Delay    11252035.0
Carrier Delay          10291572.0
NAS Delay               5573882.0
Weather Delay           2044093.0
Security Delay            50116.0
dtype: float64
Largest recorded delay cause: Late Aircraft Delay


### Business interpretation

> **[Largest cause]** contributed the most recorded delay minutes.

**Recommendation:** prioritize investigation of the largest recorded delay contributors.

Important: BTS recorded delay causes and observed weather conditions are different concepts.


## 11. Weather impact

### Business question
**Are disruptions different under bad-weather conditions?**

The statistics notebook already tested weather relationships. Here we translate those results into business language.


In [13]:
if "bad_weather" in df.columns:
    weather_summary = (
        df.groupby("bad_weather")
        .agg(
            flights=("bad_weather", "size"),
            delay_rate=("arr_del15", "mean"),
            cancellation_rate=("cancelled", "mean"),
            diversion_rate=("diverted", "mean")
        )
    )

    weather_summary["delay_rate"] *= 100
    weather_summary["cancellation_rate"] *= 100
    weather_summary["diversion_rate"] *= 100

    weather_summary.index = [
        "Normal Weather" if x == 0 else "Bad Weather"
        for x in weather_summary.index
    ]

    weather_summary
else:
    print("Column 'bad_weather' was not found.")


### Business interpretation

If bad-weather rates are higher:

> Flights operating under bad-weather conditions showed higher observed disruption levels than flights under normal weather.

**Business action:** monitor weather-sensitive airports more closely and improve disruption planning during poor-weather periods.

### Important
Use **"associated with"**, not **"caused by"**.

The analysis does not by itself prove causation.


## 12. Statistics — business interpretation

The completed statistics analysis tested:

1. Mean arrival delay vs 0
2. Weekday vs weekend departure delay
3. Cancellation vs airline
4. Arrival delay across airlines
5. Departure delay vs arrival delay
6. Cancellation vs bad weather
7. Diversion vs bad weather

The project uses **α = 0.05** and emphasizes that statistical significance is not the same as business significance. fileciteturn10file0L1-L8

Important completed results:

- Mean arrival delay ≈ **7.25 minutes**
- Weekday mean departure delay ≈ **13.12 minutes**
- Weekend mean departure delay ≈ **15.62 minutes**
- Pearson correlation between departure and arrival delay ≈ **0.9693**
- Bad weather and cancellation showed statistically significant association.
- Bad weather and diversion showed statistically significant association; the reported p-value was approximately **1.11 × 10⁻¹⁹**. fileciteturn10file3L1-L4

Because the dataset is very large, a very small p-value does not automatically mean the business effect is large. Report the actual difference or effect size too. fileciteturn10file14L1-L12


## 13. Strong business insights

Use only **5–8 strong findings** in the final project.
### Insight 1 — Overall Operational Performance

**Finding:**  
During January–March 2026, the dataset contains **1,847,242 flights**. Out of these, **1,780,194 flights were completed**. The overall arrival delay rate was **21.59%**, while **3.37% of flights were cancelled** and **0.26% were diverted**.

**Evidence:**
- **Total flights:** 1,847,242
- **Completed flights:** 1,780,194
- **Arrival delay rate:** 21.59%
- **Cancellation rate:** 3.37%
- **Diversion rate:** 0.26%

**Business impact:**  
The results show that **flight delays were a larger operational issue than cancellations or diversions** during the study period. Therefore, improving delay performance should be a major operational focus.

**Recommendation:**  
Operations teams should investigate where and when delays are concentrated, particularly by:

- Airport
- Airline
- Route
- Departure hour
- Day of week
- Recorded delay cause
- Weather condition

This can help identify the specific operational areas contributing most to disruption.



---

### Insight 2 — Departure delays carry into arrival
**Finding:** Departure and arrival delay have a very strong positive linear relationship.

**Evidence:** Pearson correlation ≈ **0.9693**.

**Business impact:** delays at departure can continue into arrival performance.

**Recommendation:** focus on preventing and reducing departure delays.

---

### Insight 3 — Weekend operations
**Finding:** Weekend average departure delay was about **15.62 minutes**, compared with **13.12 minutes** on weekdays.

**Business impact:** weekend operations may require additional attention.

**Recommendation:** review weekend scheduling, airport workload, and delay causes.

---

### Insight 4 — Airline differences
**Finding:** Airlines show different observed delay and cancellation patterns.

**Recommendation:** investigate the highest-risk airlines after considering flight volume, airport mix, routes, and delay causes.

---

### Insight 5 — Airport hotspots
**Finding:** Some airports show higher observed delay levels.

**Recommendation:** prioritize consistently underperforming airports for deeper operational review.

---

### Insight 6 — Delay drivers
**Finding:** [Largest recorded delay cause] contributed the most recorded delay minutes.

**Recommendation:** prioritize root-cause investigation in this area.

---

### Insight 7 — Weather and disruption
**Finding:** Bad-weather status was statistically associated with cancellation and diversion.

**Recommendation:** increase monitoring and disruption planning at weather-sensitive airports.

**Important:** association is not proof that weather alone caused the disruption.


## 14. Business recommendations

1. **Monitor airport hotspots** using delay rate, average delay, cancellation rate, and total delay minutes.
2. **Focus on the largest recorded delay causes** rather than treating every cause equally.
3. **Review weekend operations** because weekend departure delays were higher than weekday delays.
4. **Monitor weather-sensitive airports** during poor-weather periods.
5. **Create an operations watchlist** for high delay, cancellation, diversion, and severe-delay areas.
6. **Continue monthly monitoring** to determine whether the problems persist beyond the January–March 2026 study period.


## 15. Limitations

- The project uses public historical flight data.
- The study period is January–March 2026.
- Weather was matched using airport, date, and scheduled departure hour.
- Weather association does not prove causation.
- BTS delay-cause fields are not the same as observed weather conditions.
- Small airlines, airports, and routes can produce unstable percentages, so minimum-volume thresholds are used.
- Statistical significance can be strong in a large dataset even when the practical difference is small.


## 16. Interview-ready explanation

### What were your main business insights?

> "After cleaning and validating the airline data, I analyzed operational performance by time, airline, airport, route, delay cause, and weather. I found that departure delays were strongly associated with arrival delays, weekend departure delays were higher than weekday delays, and airline and airport performance varied. I also integrated historical weather data and found statistical evidence of an association between bad weather and cancellation/diversion. I converted these findings into operational recommendations and a Power BI dashboard."

### Did weather cause the delays?

> "I would not say weather caused the delays. My analysis shows an association between bad weather and disruption. A causal claim would require a stronger methodology that controls for other factors."

### What is the business value?

> "The project helps operations teams identify where disruption is concentrated, understand the largest recorded delay drivers, and prioritize airports, airlines, routes, and periods that need further investigation."


# ✅ BUSINESS INSIGHTS COMPLETE

### Final project flow

**Business Problem → Data Requirements → Data Collection → Data Cleaning → Data Validation → Airport Reference → Weather Integration → EDA → Statistics → Business Insights → Recommendations → Power BI**

The goal of this stage is simple:

**What happened? → Why does it matter? → What should the business investigate?**

No additional advanced statistics are required for this stage.
